### DISCUSSION TITLE
Trying to render Textbox in custom class that extends Group (V6)

### DISCUSSION THREAD

**Discussion by Naosyth** [fabricjs/fabric.js#9583] [Q&A] [answered]
Version 6.0.0-beta15

I'm trying to create a custom `GroupWithText` class which will act as a normal group, but also have an editable label. I figured the easiest way to achieve this was to additionally render a Textbox, since Textboxes already handle rendering selection/cursor/etc.

Currently, my code successfully renders the group, the group's objects, and the Textbox, but when I call `.enterEditing()` on the textbox, it does not render the editing state. If I call `this.canvas.renderAll()` after entering editing, it will render once, but won't render the blinking cursor or any of my text changes, until I exit editing, at which point the Textbox renders my newly entered text.

This is what my class currently looks like
```
import { classRegistry, Group, Textbox } from 'fabric'

export default class GroupWithText extends Group {
  static type = "GroupWithText"

  constructor(...args) {
    super(...args)
    this.test = new Textbox('test', { left: this.left, top: this.top, width: this.width, height: this.height })
  }

  enterEditing() {
    this.test.enterEditing()
    this.canvas.setActiveObject(this.test)
    this.canvas.renderAll()
  }

  // originalRender = this.render
  render(ctx) {
    super.render(ctx)

    this.test.left = this.left
    this.test.top = this.top
    this.test.render(ctx)
  }

}

classRegistry.setClass(GroupWithText)
```

Is there a better way to do this? Or at least some way to tell it to keep rendering the textbox on every frame while it's in editing mode? Ideally without also constantly re-rendering the group and all of its objects

**Comment by ShaMan123** [comment, marked as answer]
You need to add the textbox as a child of the group so it is picked up by the canvas
```typescript
  constructor(...args) {
    super(...args)
    this.test = new Textbox('test', { left: this.left, top: this.top, width: this.width, height: this.height })
   this.add(this.test)
  }
```

**Reply by ShaMan123** [reply]
Then you do not need to override rendering

**Reply by Naosyth** [reply]
I was hoping to avoid that, so that the textbox doesn't clutter the group's children. I suppose if there is no better alternative, I can deal with it, but I'd really like to keep the text separate since it doesn't make sense to be able to remove the text from a `GroupWithText` object. We also do some funky stuff with groups elsewhere in this code that I'm hoping to refactor eventually, but right now having this textbox not show up in in the group's children would make working with that old code a bit easier

**Reply by Naosyth** [reply]
I was able to get the cursor to render by tweaking my code a bit, it now looks like this
```
  constructor(...args) {
    super(...args)
    this.test = new Textbox('test', { left: this.left, top: this.top, width: this.width, height: this.height })
  }

  enterEditing() {
    this.test.canvas = this.canvas
    this.test.enterEditing()
    this.canvas.setActiveObject(this.test)
  }

  // originalRender = this.render
  render(ctx) {
    super.render(ctx)
    this.test.left = this.left
    this.test.top = this.top
    this.test.render(ctx)
  }
```
By setting `.canvas` on the Textbox, it started rendering properly. I'm fully expecting this to break in some horrible way once I start actually using this class for real, but maybe it will just magically work without any bugs 🙃
